Releasing & Versioning (#73)
bmwcarit/MoCOCrW: RELEASING.md

# Release Process

## Versioning
The versioning follows the process described at [semver.org](https://semver.org/).

### Public API
We consider the public API of this project to be whatever is included in the
`LIBRARY_PUBLIC_HEADERS` defined in `src/CMakeLists.txt`. Currently, this includes
all the header files in `src/mococrw/`. Changes to the files will be versioned according
to semantic versioning scheme.

## PRs & Releasing

### PR Checklist

> (probably to be extended)

Reviewer: Remove this comment. If we want to extend this, we will, if we don't want to, we'll have this comment sitting there that would no longer be correct.

Author: done

> (probably to be extended)

Reviewer 2: ? What is this supposed to mean?

Author: done

What should be done with every PR (if no new release is created):

> * CHANGELOG is updated with the details of the change (Unreleased section)

Reviewer 2: Rather with a summary of the change than the details? Nobody cares about (implementation) details.

Reviewer 2: Additionally: CI testing against project specific CI systems.

Author: done

### Releasing Checklist

> (probably to be extended)

Reviewer: Same here.

Author: done

> (probably to be extended)

Reviewer 2: Same here…

Author: done

Check if the following steps have been performed **before** creating a new release. This includes
all kinds of releases (Major, Minor and Patch releases)
* README is updated if necessary
* CHANGELOG is updated (add the new release)

> * Library .so version is increased (Major releases only)

Reviewer: We should only increase the SOVERSION iff the ABI changes. There is no need to increase the SOVERSION if we do not break the ABI.

Also note that the library SOVERSION is not related to the MoCOCrW release version number. There is no need to have a MoCOCrW 2.0 install a `libmococorw.so.2` file.

Consequently, I'm not sure using 3.0 as initial version number is necessary.

Author: Hmm yeah let's try to decouple that. But I guess ABI will be broken on any major release anyway. But we could skip having a major release for every ABI breaking change.

> * Library .so version is increased (Major releases only)

Reviewer 2: hint to where: `src/CMakeLists.txt`?

Author: done

> If all the steps above have been done you are good to go and create a new github release.

Reviewer: Maybe we should recommend pushing a signed annotated tag, and then promote that to a GitHub release using the GitHub UI? WDYT?

Author: sound good, done

This will also create a tag with the version number (make sure to increase the version number
according to [semver.org](https://semver.org/))

### Release Frequency

> We are developing this library based on our needs. That is, there won't be regular releases.

Reviewer: ```suggestion
We are developing this library based on our needs. That is, there is no regular release schedule.
```

"won't" is colloquial.

Author: done

Instead we are going to create releases based on added functionality.

### Release Strategy:

> In general, we won't create a new release for every PR. The following specifies what

Reviewer: ```suggestion
In general, we will not create a new release for every PR. The following specifies what
```

Author: done
